compiler/qsc_frontend/src/typeck/infer.rs

```diff
@@ -1018,27 +1052,32 @@ fn check_ctl(op: Ty, with_ctls: Ty, span: Span) -> (Vec<Constraint>, Vec<Error>)
     };
 
     let qubit_array = Ty::Array(Box::new(Ty::Prim(Prim::Qubit)));
-    let ctl_input = Box::new(Ty::Tuple(vec![qubit_array, *arrow.input]));
-    (
+    let ctl_input = RefCell::new(Ty::Tuple(vec![qubit_array, arrow.input.borrow().clone()]));
+    let ret = (
         vec![
             Constraint::Superset {
                 expected: FunctorSetValue::Ctl,
-                actual: arrow.functors,
+                actual: *arrow.functors.borrow(),
                 span,
             },
             Constraint::Eq {
-                expected: Ty::Arrow(Box::new(Arrow {
+                expected: Ty::Arrow(Rc::new(Arrow {
                     kind: arrow.kind,
                     input: ctl_input,
-                    output: arrow.output,
-                    functors: arrow.functors,
+                    output: arrow.output.clone(),
+                    functors: arrow.functors.clone(),
                 })),
                 actual: with_ctls,
                 span,
             },
         ],
         Vec::new(),
-    )
+    );
+    #[allow(
+        clippy::let_and_return,
+        reason = "rust requires the explicit let-biding to satisfy lifetime guarantees"
```

Comment:
```suggestion
        reason = "rust requires the explicit let-binding to satisfy lifetime guarantees"
```

Reply:
Actually, you don't even need the clippy suppression if you just pull the `*arrow` access out of this expression, and return the expression directly:

```rust
    let actual = *arrow.functors.borrow();
    (
        vec![
            Constraint::Superset {
                expected: FunctorSetValue::Ctl,
                actual,
                span,
            },
            Constraint::Eq {
                expected: Ty::Arrow(Rc::new(Arrow {
                    kind: arrow.kind,
                    input: ctl_input,
                    output: arrow.output.clone(),
                    functors: arrow.functors.clone(),
                })),
                actual: with_ctls,
                span,
            },
        ],
        Vec::new(),
    )
```

Reply:
Good call, that is a nicer solution.